tensorrt_llm/_torch/pyexecutor/_util.py

```diff
@@ -34,281 +34,331 @@
+
+        available_kv_mem = (total_gpu_memory - peak_memory +
+                            alloc_kv_tokens * kv_size_per_token) * fraction
+        logger.info(
+            f"Peak memory during memory usage profiling (torch + non-torch): {peak_memory / (GB):.2f} GiB, "
+            f"available KV cache memory when calculating max tokens: {available_kv_mem / (GB):.2f} GiB, "
+            f"fraction is set {fraction}, kv size is {kv_size_per_token}")
+        max_tokens = int((available_kv_mem) // kv_size_per_token)
+        max_tokens = max(max_tokens, 0)
+        return max_tokens
+
+    def _create_dummy_context_requests(
+            self, input_seq_len: int) -> List[trtllm.Request]:
+        vocab_size = self._model_engine.model.model_config.pretrained_config.vocab_size
+        max_num_tokens = self._executor_config.max_num_tokens
+
+        requests = []
+        input_seq_len = min(max_num_tokens, input_seq_len)
+        remaining_tokens = max_num_tokens
+        while remaining_tokens > 0:
+            input_seq_len = min(input_seq_len, remaining_tokens)
+            input_tokens = [
+                random.randint(0, vocab_size - 1) for _ in range(input_seq_len)
+            ]
+            request = trtllm.Request(input_tokens,
+                                     max_tokens=1,
+                                     streaming=False,
+                                     sampling_config=trtllm.SamplingConfig(),
+                                     output_config=trtllm.OutputConfig(),
+                                     end_id=-1)
+            requests.append(request)
+            remaining_tokens -= input_seq_len
+        return requests
+
+    def _get_token_num_for_estimation(self) -> int:
+        """Compute KV cache capacity required for estimate_max_kv_cache_tokens to succeed."""
+        executor_config = self._executor_config
+        if 'cp_type' in self._mapping.cp_config:
+            raise ValueError(
+                "KV cache size estimation not supported with context parallelism."
+            )
+        # estimate_max_kv_cache_tokens submits self._dummy_reqs
+        num_cache_blocks = 0
+        num_extra_tokens_per_seq = 1  # account for generated tokens
+        spec_cfg = executor_config.speculative_config
+        if spec_cfg is not None:
+            num_extra_tokens_per_seq += spec_cfg.max_draft_tokens
+            num_extra_tokens_per_seq += spec_cfg.num_extra_kv_tokens
+        for req in self._dummy_reqs:
+            num_req_tokens = len(req.input_token_ids) + num_extra_tokens_per_seq
+            # Requests cannot share KV cache blocks. Round up to nearest integer multiple of block size.
+            num_cache_blocks += (num_req_tokens +
+                                 executor_config.tokens_per_block -
+                                 1) // executor_config.tokens_per_block
+        return num_cache_blocks * executor_config.tokens_per_block
+
+    def maybe_prepare_estimation(self) -> bool:
```

Comment:
`maybe` is misleading in this method's name. How about `is_estimating_kv_cache` instead.